### ISSUE TITLE
Unsoundness with "exists" variable binding

### ISSUE THREAD

**Issue by mrkmarron** [Z3Prover/z3#5053]
Code below -- returns sat but formula is unsatisfiable.

Key clause is `(exists ((@n BNat)) (and (bvult @n (_ bv2 4)) (= (_ bv0 4) (@@get @la @n))))` where the `@@get` operation is only well defined when `@n < 2`. Z3 seems to be ignoring the constraint on `@n` and then generating an arbitrary result for the get which it can use to satisfy the exists. 

Interestingly I tried using explicit enumeration with an `ite` guard on `@n` which is also ignored. Replacing the `@n` in the body with the literal produces the correct result as well as explicitly blasting the existential into an `or` clause. Commented out below. 

```
(set-option :smt.mbqi true)
(set-option :timeout 10000)

;;
;; Define sort aliases for Int/Nat/BigInt/BigNat/Float/Decimal/Rational/String representation options
;;
(define-sort BInt () (_ BitVec 4))
(define-sort BNat () (_ BitVec 4))

;;
;; Any Concept datatypes
;;
(declare-datatypes (
    (nscore__list_nscore__int_I2I$uli 0)
    (nscore__list_nscore__int_I2I 0)
    ) (
    ( (@@cons_nscore__list_nscore__int_I2I_empty ) (@@cons_nscore__list_nscore__int_I2I_slice (@@cons_nscore__list_nscore__int_I2I_slice_l nscore__list_nscore__int_I2I) (@@cons_nscore__list_nscore__int_I2I_slice_start BNat) (@@cons_nscore__list_nscore__int_I2I_slice_end BNat)) (@@cons_nscore__list_nscore__int_I2I_concat2 (@@cons_nscore__list_nscore__int_I2I_concat2_l1 nscore__list_nscore__int_I2I) (@@cons_nscore__list_nscore__int_I2I_concat2_l2 nscore__list_nscore__int_I2I)) (@@cons_nscore__list_nscore__int_I2I__1 (@@cons_nscore__list_nscore__int_I2I__1_idx0 BInt)) )
    ( (nscore__list_nscore__int_I2I@cons (nscore__list_nscore__int_I2I@size BNat) (nscore__list_nscore__int_I2I@uli nscore__list_nscore__int_I2I$uli)) )
))

(define-fun-rec @@get ((l nscore__list_nscore__int_I2I) (n BNat)) BInt
  (let ((@list_contents (nscore__list_nscore__int_I2I@uli l)))
    (ite (is-@@cons_nscore__list_nscore__int_I2I_slice @list_contents)
      (@@get (@@cons_nscore__list_nscore__int_I2I_slice_l @list_contents) (bvadd n (@@cons_nscore__list_nscore__int_I2I_slice_start @list_contents)))
      (ite (is-@@cons_nscore__list_nscore__int_I2I_concat2 @list_contents)
        (let ((@l1 (@@cons_nscore__list_nscore__int_I2I_concat2_l1 @list_contents)))
          (let ((@l1size (nscore__list_nscore__int_I2I@size @l1)))
            (ite (bvult n @l1size)
              (@@get @l1 n)
              (@@get (@@cons_nscore__list_nscore__int_I2I_concat2_l2 @list_contents) (bvsub n @l1size))
            )
          )
        )
        (@@cons_nscore__list_nscore__int_I2I__1_idx0 @list_contents)
      )
    )
  )
)

(declare-const @l1 nscore__list_nscore__int_I2I)
(assert (= @l1 (nscore__list_nscore__int_I2I@cons (_ bv1 4) (@@cons_nscore__list_nscore__int_I2I__1 (_ bv1 4)))))

(declare-const @l2 nscore__list_nscore__int_I2I)
(assert (= @l2 (nscore__list_nscore__int_I2I@cons (_ bv1 4) (@@cons_nscore__list_nscore__int_I2I__1 (_ bv1 4)))))

(declare-const @la nscore__list_nscore__int_I2I)
(assert (= @la (nscore__list_nscore__int_I2I@cons (_ bv2 4) (@@cons_nscore__list_nscore__int_I2I_concat2 @l1 @l2))))

(declare-const @has Bool)
(assert (= @has (exists ((@n BNat)) (and (bvult @n (_ bv2 4)) (= (_ bv0 4) (@@get @la @n))))))

;;(assert (= @has (exists ((@n BNat)) 
;;  (ite (= @n (_ bv0 4))
;;    (= (_ bv0 4) (@@get @la @n))
;;    (ite (= @n (_ bv1 4))
;;      (= (_ bv0 4) (@@get @la @n))
;;      false
;;    )
;;  )
;;)))

(declare-const @g1 BNat)
(assert (= @g1 (@@get @la (_ bv0 4))))

(declare-const @g2 BNat)
(assert (= @g2 (@@get @la (_ bv1 4))))

;;(assert (= @has (or (= (_ bv0 4) (@@get @la (_ bv0 4))) (= (_ bv0 4) (@@get @la (_ bv1 4))))))

;;(assert (= @has (exists ((@n BNat)) 
;;  (ite (= @n (_ bv0 4))
;;    (= (_ bv0 4) @g1)
;;    (ite (= @n (_ bv1 4))
;;      (= (_ bv0 4) @g2)
;;      false
;;    )
;;  )
;;)))

(assert @has)
(check-sat)
(get-model)
```

**Comment by NikolajBjorner** [comment]
Looks like your bug was already fixed as reported by #5016.

**Comment by mrkmarron** [comment]
Yes, you are right. The nightly build works correctly. Thanks!
